### ISSUE TITLE
🚀 Build & deployment optimizations [tracker]

### ISSUE THREAD

**Issue by andrewconnell** [pnp/blog#637]
Tracker issue for build & deployment optimizations. The goal of this project is to simplify & speed up builds & deployment tasks

- [x] #638
- [x] #639

**Comment by andrewconnell** [comment]
- consolidated from 2 workflows used in build & deployment
- before, builds & deployments took 20-45 minutes...

    ![Screen Shot 2022-10-16 at 8 03 15 AM](https://user-images.githubusercontent.com/2068657/196034370-831e4fef-db0e-479c-85bf-54f2f1c472ea.png)

- now builds take 10 minutes (using caches) - 20 minutes (on the first run of the month to build caches)
  - the majority of the build+deployment is based on the deployment... *it takes ~10m to copy 1.7GB & 9,000+ files 🤷‍♂️*

  ![Screen Shot 2022-10-16 at 8 29 23 AM](https://user-images.githubusercontent.com/2068657/196035457-33613455-5f75-4486-a7b3-6f7d43bb10f1.png)

Changes applied... more information on what changed is found below...

# TL;DR

## Previously... 

Previously, the automated build (defined in the [workflow **gh-pages.yml**](https://github.com/pnp/blog/blob/30786a99b0d1108e944598e7e70c98e7d61a1a64/.github/workflows/gh-pages.yml)) would do the following on every production build > deploy:

- full rebuild of the site
- using the action [peaceiris/actions-gh-pages@v3](https://github.com/marketplace/actions/github-pages-action)...
  - add the build images to the branch **gh-pages**
  - trigger a child workflow that would checkout the the files on the **gh-pages** repo
  - deploy them to the GitHub Pages configuration for this site

This resulted in two workflows that, combined, would take from 20-45 minutes...


That workflow (gh-pages.yml) has been replaced by the workflow in the commit associated with the subtasks (#638 & #639) of this task.

## Now... (ie: what changed?)

The dependency on [peaceiris/actions-gh-pages@v3](https://github.com/marketplace/actions/github-pages-action) was removed as unnecessary.

### WRT the build of the site...

Now, the workflow will build the site & use, if present, the previously Hugo-generated CSS & images cached in the **./resources** folder. The generation of these files can take 20+ minutes...

![Screen Shot 2022-10-16 at 8 27 56 AM](https://user-images.githubusercontent.com/2068657/196035425-403ffa8b-e357-40b5-81ee-1a124b302d98.png)

... but subsequent builds that leverage cache will be as fast as 20 ***seconds***... but always less than 1 minute...

![Screen Shot 2022-10-16 at 8 22 41 AM](https://user-images.githubusercontent.com/2068657/196035195-a1c848c4-ef45-41e6-98d4-ded41c92dcea.png)

> NOTE: The generated cache is regenerated monthly (first run of the month). All subsequent builds will only have to generate images dynamically if they aren't already in the cache (ie: changes to existing images or new images added to the site). This will only modestly increase the build time over the baseline ~20 seconds. By using the cache, we're avoiding Hugo having to process ~5,600 images on each build (at the time this work was done).
>
> NOTE: Force the expiry of the cache using the manually run workflow **[Purge repo GitHub Actions caches](https://github.com/pnp/blog/blob/main/.github/workflows/purge-repo-caches.yml)**

These changes eliminate about 10-20m of build time on every build by leveraging previously Hugo-generated images thanks to GitHub Action's caching.

### WRT the deployment of the site...

Once the site is built, the generated artifacts in **./public** are saved as an artifact. The workflow then deploys the generated files using the artifact directly to the GitHub pages.

We're no longer using the **gh-pages** branch as a deployment vehicle.

**Comment by andrewconnell** [comment]
BTW, builds are showing lots of annotations about the `set-output` command deprecation. This coming from an 1st-party GH action we're using & stems from [this announcement](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/)

We can't control that, but GH will update the action so those go away soon. They've already addressed it on numerous other 1st-party actions and will presumably leverage it on this one as well.
